backend/assignments/2/src/main/java/com/caching/services/GeoServices.java

```diff
@@ -0,0 +1,139 @@
+        String region = results.get("region").asText();
+
+        return new ResponseDTO(address,latitude,longitude,region);
+    }
+
+    /**
+     * Processing geocoding by given address and apply/use caching if required.
+     * @param address
+     * @return responseDTO which consists of latitude and longitude which we got from caching
+     * along with address and region
+     */
+    @Cacheable(value = "geocoding", key = "#address", unless = "T(com.caching.services.GeoServices).isExcludedRegion(#result.region)")
+    public ResponseDTO getLatLong(String address){
+        String apiUrl = "http://api.positionstack.com/v1/forward?access_key=".concat(accessKey).concat("&query=").concat(address);
+        JsonNode jsonNode = restTemplate.getForObject(apiUrl,JsonNode.class);
+        if(jsonNode == null){
+            logger.error(Constants.INVALID);
+            throw new InvalidRequestException(Constants.INVALID);
+        }
+
+        JsonNode results = jsonNode.get("data").get(0);
+        if(logger.isInfoEnabled() && results != null){
+            logger.info("Result of address ".concat(address).concat(" is fetched from API!"));
+            logger.info("Latitude : ".concat(results.get("latitude").asText()).concat(" Longitude : ")
+                    .concat(results.get("longitude").asText()));
+            return jsonToResponseDTO(results);
+        }
+        else{
+            logger.error("Invalid address!");
+            throw new InvalidRequestException("Invalid address!");
+        }
+    }
+
+    /**
+     * Processing reverse-geocoding by given latitude and longitude and apply/use caching if possible
+     * @param latitudeStr
+     * @param longitudeStr
+     * @return responseDTO which consists of address and region which we got from caching
+     * along with latitude and longitude
+     */
+    @Cacheable(value = "reverse-geocoding", key = "{#latitude,#longitude}", unless = "T(com.caching.services.GeoServices).isExcludedRegion(#result.region)")
+    public ResponseDTO getAddress(String latitudeStr, String longitudeStr){
+        if(isNotValid(latitudeStr,longitudeStr)){
+            logger.error("Invalid latitude or longitude!");
+            throw new InvalidRequestException("Invalid latitude or longitude!");
+        }
+        double latitude = Double.parseDouble(latitudeStr);
+        double longitude = Double.parseDouble(longitudeStr);
+
+        String apiUrl = "http://api.positionstack.com/v1/reverse?access_key=".concat(accessKey).concat("&query=")
+                .concat(Double.toString(latitude)).concat(",").concat(Double.toString(longitude));
+        JsonNode jsonNode = restTemplate.getForObject(apiUrl,JsonNode.class);
+        if(jsonNode == null){
+            logger.error(Constants.INVALID);
+            throw new InvalidRequestException(Constants.INVALID);
+        }
+
+        JsonNode results = jsonNode.get("data").get(0);
+        if(logger.isInfoEnabled() && results != null){
```

Comment:
Simplify your logger mechanism. Why do we have to check every time the enabling/disabling for simple logs.